### PR TITLE
C22-Pheonix - Amber Edwards

### PR DESCRIPTION
**Pull request by Msambere**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+
+const ChatEntry = (props) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: `ChatEntry` an d`ChatLog` use a single `props` param, but your optional components use a destructured object. Personally, I prefer the destructured style, since it makes the expected component attributes a bit more clear. And it's fine to use a mixture of styles in this project, but in general, try to pick one style or the other.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={`chat-entry ${localRemote}`}>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p className={senderColor}>{props.body}</p>
+        <p className="entry-time">
+          <TimeStamp time = {props.timeStamp}/>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:20]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Nit: no spaces around attribute `=`

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,40 @@
...
-  // Fill with correct proptypes
+  local: PropTypes.bool.isRequired,
+  color: PropTypes.string.isRequired,
+  isLiked: PropTypes.bool,
+  sender: PropTypes.string.isRequired,
+  id: PropTypes.number,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool,
+  onLikedToggle: PropTypes.func.isRequired
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:37]: 👀 The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) marked them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: dist/assets/index-BFUh5if4.css
Comment (anselrognlie) [inline dist/assets/index-BFUh5if4.css:1]: It should be possible to add `dist` to the `.gitignore` so the `dist` folder doesn't get checked in to your code branch. Build products generally shouldn't get checked into the source repo.

File: package.json [hunk 4]
```diff
@@ -2,13 +2,16 @@
   "name": "react-chatlog",
   "private": true,
   "version": "0.0.0",
+  "homepage": "https://msambere.github.io/react-chatlog",
```
Comment (anselrognlie) [inline package.json:5]: Since you set the `base` in `vite.config.js` it shouldn't be necessary to set `homepage` here. I'm not sure what effect if any this has for a vite build.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  return(
+    <main>
+      {chatEntries}
+    </main>
+  );
+};
+
+ChatLog.propTypes ={
+  onLikedToggle: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:29]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  entries: PropTypes.arrayOf(PropTypes.shape({ // You can create a constant that has this object shape and import it
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:30]: Great observation that we could define this shape elsewhere and import it. This can reduce the visual noise in the prop types, especially if the same shape is used in multiple places. But even for a single instance, it can be helpful to give the shape an actual name.

The key observation is that the code defining the propTypes is just that, code. So we can use regular JS techniques to reduce redundancy.

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) =>{
+  const chatEntries = props.entries.map((entry)=>{
+    return(
+      <ChatEntry
+        key={entry.id}
+        id ={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:9]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) =>{
+  const chatEntries = props.entries.map((entry)=>{
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:5]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/components/ChatLog.jsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        body={entry.body}
+        onLikedToggle = {props.onLikedToggle}
+        liked ={entry.liked}
+        local={entry.local}
+        color={entry.color}
+      />
+    );
+  });
+
+  return(
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:21]: Nit: leave a space between `return` and `(`. `return` isn't a function.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,101 @@
...
+  const [messageData, setMessageData] = useState(MESSAGES);
+
+  const findChatMembers = (messages) =>{
+    let chatMembers =[];
+    for(let message of messages){
+      if (!chatMembers.includes(message.sender)){
+        chatMembers.push(message.sender);
+      };
+    };
+    // Beenish suggestion: [...new Set(Data.map(message => message.sender))]
```
Comment (anselrognlie) [inline src/App.jsx:21]: 👍 Using `map` to extract the sender information then uniquing that with a `Set` would work great!

While we don't expect the members in the chat to be large, from a purely theoretical perspective the implemented `includes` approach is quadratic on the number of messages (assume ever message has a unique sender) while the set approach is linear. Does it matter here? Probably not. But it's worth keeping an eye out for cases like this anyway.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,101 @@
...
+    };
+    chatMembers= chatMembers.map((member,i) =>{
+      return {...member, id:i};
+    });
+    return chatMembers;
+  };
+
+  const memberColorInfo = getMemberColorInfo(messageData);
+
+  const toggleLiked = (messageID)=>{
```
Comment (anselrognlie) [inline src/App.jsx:57]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

Nit: leave  spaces around `=>`

File: src/App.jsx [hunk 12]
```diff
@@ -1,14 +1,101 @@
...
+    setMessageData(messages => {
```
Comment (anselrognlie) [inline src/App.jsx:58]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

File: src/App.jsx [hunk 13]
```diff
@@ -1,14 +1,101 @@
...
+      return messages.map(message => {
```
Comment (anselrognlie) [inline src/App.jsx:59]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 14]
```diff
@@ -1,14 +1,101 @@
...
+        if (message.id === messageID) {
+          return { ...message, liked: !message.liked };
```
Comment (anselrognlie) [inline src/App.jsx:61]: We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 15]
```diff
@@ -1,14 +1,101 @@
...
+        if (message.sender === name) {
+          return {...message, color: newColor};
+        }else{
+          return message;
+        }
+      });
+    });
+  };
+
+  const totalLikes = messageData.filter((entry)=> entry.liked === true).length;
```
Comment (anselrognlie) [inline src/App.jsx:81]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Great idea to filter down the data to the liked messages and use the list length as the count. This is a really understandable alternative to the more canonical `reduce` approach. Compared to `reduce`, it does make a list potentially as long as the message list, but since we know a similar copying gets done anyway during a render (building the `ChatEntry` components) on the whole it's not any more costly.

Comment (anselrognlie) [reply src/App.jsx:81]: Since `liked` is already a boolean, it should be sufficient to return it directly
```js
  const totalLikes = messageData.filter(entry => entry.liked).length;
```

File: src/components/ChatEntry.jsx [hunk 16]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+
+const ChatEntry = (props) => {
+  const heartColor = props.liked ? '❤️':'🤍' ;
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

Nit: spacing
```js
  const heartColor = props.liked ? '❤️' : '🤍';
```

File: src/components/ChatEntry.jsx [hunk 17]
```diff
@@ -1,20 +1,40 @@
...
+  const localRemote = props.local == true ? 'local': 'remote';
+  const senderColor = props.color;
+
+  const handleLikedClicked = () => {
+    props.onLikedToggle(props.id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:11]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 18]
```diff
@@ -1,20 +1,40 @@
...
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p className={senderColor}>{props.body}</p>
+        <p className="entry-time">
+          <TimeStamp time = {props.timeStamp}/>
+        </p>
+        <button onClick={handleLikedClicked}className="like">{heartColor}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:22]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

Nit: spacing
```html
        <button onClick={handleLikedClicked} className="like">{heartColor}</button>
```

File: src/App.jsx [hunk 19]
```diff
@@ -1,14 +1,101 @@
 import './App.css';
+import ColorChoices from './components/ColorChoices';
+import ChatLog from './components/ChatLog';
+import OG_MESSAGES from './data/messages.json';
+import { useState } from 'react';
+
+const MESSAGES = OG_MESSAGES.map( message =>{
+  return {...message, color: '', local: message.sender === OG_MESSAGES[0].sender};
```
Comment (anselrognlie) [inline src/App.jsx:8]: Including the color and locality of each message allows su to get this information down to the `ChatEntry` where it needs to be used to affect the rendering, but it is a bit redundant. But there are defeinitely cases where this might be the prefered approach.

Here, I consider the color and locality to be more a property of a sender than the message. That is, by knowing the sender, we know which values apply to any message sent by that sender. So one way we could address this is in addition to the message data, we could include a dictionary of sender -> config data, so that as a `ChatEntry` renders itself, it can look up the color and locality of the sender.

If we wanted to retain a `ChatEntry` that's less tied to those specific extra data values, we could introduce an additional component layer that represents a "sender-aware" chat entry, which extracts the sender-related data, but then passes that down to the "raw" chat entry that is just explicitly told the color and locality.

File: src/App.jsx [hunk 20]
```diff
@@ -1,14 +1,101 @@
...
+    for(let message of messages){
+      if (!chatMembers.includes(message.sender)){
+        chatMembers.push(message.sender);
+      };
+    };
+    // Beenish suggestion: [...new Set(Data.map(message => message.sender))]
+    return chatMembers;
+  };
+
+  const chatTitle = (members) =>{
```
Comment (anselrognlie) [inline src/App.jsx:25]: Nice helper to encapsulate the logic of building an appropriate title based on the number of chat participants. String formatting for displays can be surprisingly finicky!

File: src/App.jsx [hunk 21]
```diff
@@ -1,14 +1,101 @@
...
+    if (members.length == 2) {
+      return `${members[0]} and ${members[1]}`;
+    }else if (members.length > 2) {
+      let lastMember = members.at(-1);
+      let title= members.slice(0,-1).join(', ');
+      title += `, and ${lastMember}`;
+      return title;
```
Comment (anselrognlie) [inline src/App.jsx:32]: Or to avoid the lone string concatenation.
```js
      const lastMember = members.at(-1);
      const titlePieces = members.slice(0, -1)
      titlePieces.push(`and ${lastMember}`);
      return titlePieces.join(', ');
```

File: src/App.jsx [hunk 22]
```diff
@@ -1,14 +1,101 @@
...
+      };
+    };
+    // Beenish suggestion: [...new Set(Data.map(message => message.sender))]
+    return chatMembers;
+  };
+
+  const chatTitle = (members) =>{
+    if (members.length == 2) {
+      return `${members[0]} and ${members[1]}`;
+    }else if (members.length > 2) {
```
Comment (anselrognlie) [inline src/App.jsx:28]: Nit: spacing before `else`
```js
    } else if (members.length > 2) {
```

File: src/App.jsx [hunk 23]
```diff
@@ -1,14 +1,101 @@
...
+      let lastMember = members.at(-1);
+      let title= members.slice(0,-1).join(', ');
+      title += `, and ${lastMember}`;
+      return title;
+    }
```
Comment (anselrognlie) [inline src/App.jsx:33]: 👀 We should include a final `else` to handle the 0 and 1 participant case. Currently, this will implicitly return `undefined` but since we're returning from some paths, we should return from all. It might be worth pulling the "Chat Between" portion of the title into this function as well to give us control over that part of the title here as well.

If we're not expecting that the 0 or 1 case could actually happen, consider returning `''` as just a fallback that we don't expect to use, but I'd still rather see a case that accounts for them.

File: src/components/ColorChoices.jsx [hunk 24]
```diff
@@ -0,0 +1,32 @@
+import ColorPicker from './ColorPicker';
+import './ColorChoices.css'
+import PropTypes from 'prop-types';
+
+
+const ColorChoices = ({changeSenderColor, memberColorInfo}) =>{
+  const colorPickers = memberColorInfo.map( member=>{
+    return(
+      <ColorPicker
+        key={member.id}
```
Comment (anselrognlie) [inline src/components/ColorChoices.jsx:10]: The key doesn't have to be a number. It just needs to be unique. So since you built the member data in a way that guarantees each member appears only once, we could use the member name as the key. This would obviate the need for grafting in the `id` value of the member data.

Of course, in a full messaging system, we would probably have some sort of id to identify the members, in which case using that would be prefered.

File: src/components/ColorPicker.jsx [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+  return(
+    <section className="color-picker">
+      <h3 className={member.color}>{member.name}&apos;s Color:</h3>
+      <section className="color-options">
+        <button onClick={(e)=>handleColorChange(e)} id='red' className="color-option">🔴</button>
+        <button onClick={(e)=>handleColorChange(e)} id='orange' className="color-option">🟠</button>
+        <button onClick={(e)=>handleColorChange(e)} id='yellow' className="color-option">🟡</button>
+        <button onClick={(e)=>handleColorChange(e)} id='green' className="color-option">🟢</button>
+        <button onClick={(e)=>handleColorChange(e)} id='blue' className="color-option">🔵</button>
+        <button onClick={(e)=>handleColorChange(e)} id='purple' className="color-option">🟣</button>
```
Comment (anselrognlie) [inline src/components/ColorPicker.jsx:16]: Since we know there's going to be more than one `ColorPicker` on a page, we should avoid using `id` in this way, since they will not be unique. There are a few ways we could approach this.
1. We could make a `ColorButton` component and pass the color value we'd like reported back as one of it's props. Just like how we can hold onto a message id in `ChatEntry` to pass that back up to a callback when the like is clicked, the `ColorButton` could pass that value back up to us when it gets clicked. This would be the most generalizable approach for most situations.
2. We could use a non-unqiue attribue on the button. Buttons can have a `value`, which we can read in the event handler. If we need to store multiple pieces of data, we can use custom [`data-` attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*) as well.
3. In the anonymous click handler function, we could code the color value to pass to the handler rather than forwarding along the event.
```html
        <button onClick={() => handleColorChange('red')} className="color-option">🔴</button>
```
or
```html
        <button onClick={() => changeSenderColor(member.name, 'red')} className="color-option">🔴</button>
```

Comment (anselrognlie) [reply src/components/ColorPicker.jsx:16]: Rather than hard coding all the buttons, consider defining a data structure with the information needed for the buttons, then iterate over it to generate the JSX.
```js
// earlier
const availColors = [['🔴', 'red'], ['🟠', 'orange'], ...];
```
then for the buttons
```html
        {
          availColors.map(colorData => (
            <button
              key={colorData[1]}
              onClick={() => handleColorChange(colorData[1])}
              className="color-option">
              {colorData[0]}
            </button>
          ))
        }
```

File: src/components/ColorPicker.jsx [hunk 26]
```diff
@@ -0,0 +1,32 @@
+import PropTypes from 'prop-types';
+import './ColorPicker.css';
+
+const ColorPicker = ({member, changeSenderColor}) =>{
+  const handleColorChange = (e) => changeSenderColor(member.name, e.target.id);
```
Comment (anselrognlie) [inline src/components/ColorPicker.jsx:5]: Rather than receiving the raw event, we could have the caller pass a paticular color
```js
  const handleColorChange = (color) => changeSenderColor(member.name, color);
```

File: src/components/ColorChoices.jsx [hunk 27]
```diff
@@ -0,0 +1,32 @@
+import ColorPicker from './ColorPicker';
+import './ColorChoices.css'
+import PropTypes from 'prop-types';
+
+
+const ColorChoices = ({changeSenderColor, memberColorInfo}) =>{
+  const colorPickers = memberColorInfo.map( member=>{
```
Comment (anselrognlie) [inline src/components/ColorChoices.jsx:7]: Great job supporting multiple color pickers even if there were more than two particpants in the chat.

File: src/App.jsx [hunk 28]
```diff
@@ -1,14 +1,101 @@
...
+    let chatMembers = [];
+    let uniqueSenders = new Set();
+    for(let message of messages){
+      if (!uniqueSenders.has(message.sender)){
+        let member = {
+          name: message.sender,
+          color: message.color,
+        };
+        chatMembers.push(member);
+        uniqueSenders.add(message.sender);
```
Comment (anselrognlie) [inline src/App.jsx:46]: Since we're treating the sender name as being unique, we could use that later on as the React key where needed, removing the need to fabricate an id below.

File: src/App.jsx [hunk 29]
```diff
@@ -1,14 +1,101 @@
...
+      let title= members.slice(0,-1).join(', ');
+      title += `, and ${lastMember}`;
+      return title;
+    }
+  };
+
+  const getMemberColorInfo = (messages) =>{
+    let chatMembers = [];
+    let uniqueSenders = new Set();
+    for(let message of messages){
```
Comment (anselrognlie) [inline src/App.jsx:39]: While in principle, this is similar to the idea that id we store the relevant data in the messages, we can extract it later and avoid additional state, it is a bit different. While the like data is truly per-message information, the color is actually a property of the sender. So while the only way that we have to get the unique sender information is from the messages, we could use the resulting unique sender information to track color information at the sender level rather than the message level (likewise for the locality) and avoid the redundancy in the messages.

In a more complete chat application, we would probably have received information about the participants in the conversation rather than needing to extract it like this.

File: src/App.jsx (hunk 19 above)
Comment (anselrognlie) [reply src/App.jsx:8]: 👍 We didn't receive any explicit data indicating which of the senders is local and which is remote, so deciding to treat the first sender as local is a reasonable decision.

File: src/App.jsx [hunk 30]
```diff
@@ -1,14 +1,101 @@
 import './App.css';
+import ColorChoices from './components/ColorChoices';
+import ChatLog from './components/ChatLog';
+import OG_MESSAGES from './data/messages.json';
+import { useState } from 'react';
+
+const MESSAGES = OG_MESSAGES.map( message =>{
+  return {...message, color: '', local: message.sender === OG_MESSAGES[0].sender};
+});
```
Comment (anselrognlie) [inline src/App.jsx:9]: Note that we could avoid the global reference in the map expression by using the optional third param passed to the `map` callback function.
```js
const MESSAGES = OG_MESSAGES.map((message, _, arr) => {
  return {...message, color: '', local: message.sender === arr[0].sender};
});
```

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
